### ISSUE TITLE
Encrypted attachments are not seekable

### ISSUE THREAD

**Issue by elevine** [couchbase/couchbase-lite-net#621]
When encrypted attachments are retrieved as a stream, the stream is not seekable.  As a result, we (@APXLabs) cannot playback encrypted video files.  To achieve this with the previous release of CBL, we made a fork that encrypted attachments with AES in CTR mode.  This approach has been working out well so far, but ideally Couchbase Lite would offer the ability to use different encryption strategies for attachments to support use cases like this.

**Comment by borrrden** [comment]
This might become a cross platform compatibility issue, and at the very least an API issue so I will have to talk to the architect about this.

**Comment by snej** [comment]
How often do you need to seek? If it's just occasionally, like to let the user jump to a different time, you can fake it — seek forward by reading and discarding data until you get to the desired spot, and seek backwards by opening a new stream and then seeking forwards from 0. Obviously the performance won't be great, but if it's not happening often it may be acceptable.

Having two different cipher modes in use would present problems determining which one is in use for which file. It's not rocket science, but from past experience I can predict that it would create a lot of headaches, especially with database upgrading and cross-platform compatibility of database files.

**Comment by Ariemeth** [comment]
Our use case is playing stored videos.

**Comment by snej** [comment]
Yes, I know that. I'm not an expert on video codecs, but my assumption is that just playing a video straight through doesn't require seeking the stream. So I'm asking what use cases involve seeking and how often, and how fast the seeks need to be.

**Comment by davidlongest** [comment]
@snej The reason for the need to seek without reading (and decrypting) is that some video codecs have the codec information stored near the end of the file. For example, some h.264 files will have the MOOV atom at the end of the file. This doesn't include when the user might seek the video.

On medium-low performance devices this can really take a long time as well as use up battery life unnecessarily on any mobile devices.

**Comment by snej** [comment]
Hm, that does sound like a problem. If I were just implementing attachment encryption now, I'd go with CTR mode; but as I said, there would now be compatibility problems. For now I think you'll need to keep using your fork, but if we get more requests from developers to make attachment streams seekable, we could revisit this in the future.

**Comment by elevine** [comment]
@snej So far, we've only made a small change that adds an `ISymmetricKey` interface which `SymmetricKey` implements.  Then, we changed all references over to using `ISymmetricKey`.  This allows for the flexibility of using different encryption for attachments, but its a small impact to the codebase.  In our application, we will have the implementation which uses AES CTR and set CBL to use it.  

Would you be willing to consider adopting that change? 

You can see the code on this small PR https://github.com/APXLabs/couchbase-lite-net/issues/3.  We  wanted to work off of the last stable release, which is why it isn't based off of your current master.  If you're open to it, I can work on submitting that PR to you.

**Comment by borrrden** [comment]
Just to be clear it's not just problems with the .NET API, but also maintaining compatibility with Java and Objective-C.  Also changing any public API like this is a breaking change that needs to wait until a major version change to obey semantic version rules, not matter how small.

**Comment by elevine** [comment]
@borrrden Makes sense w/r to maintaining API compatibility with the other languages.  We don't like having a fork, so getting this on your radar and knowing it would get into CBL at some point in the (hopefully near) future would be helpful.

Wouldn't this be a minor version change according to semver, since its a backwards compatible?

**Comment by snej** [comment]
That's a good idea — adding a per-database configuration to specify encryption modes. (With the 1.2 API we could even add this as a property to the DatabaseOptions object.) That way the compatibility issues from changing modes become the app developer's responsibility :)

I'll record this in our list of API proposals.

**Comment by snej** [comment]
OK, done. BTW, just to make sure: do you guys _need_ the media files to be encrypted? Or is it just a side effect of encrypting the documents themselves? I'm asking because another API option is to allow attachment encryption to be configured independently of document encryption.

**Comment by elevine** [comment]
Our requirement is to encrypt all of our data at rest, so we do need the media files to be encrypted.  I like the idea of having more flexibility, so configuring them separately sounds great!

**Comment by borrrden** [comment]
@elevine  Source wise it is compatible, but it breaks binary compatibility because the method signature changes (at least I think so).  This is not a big deal for anyone consuming couchbase lite directly, but it would break any libs out there that have a nuget dependency on couchbase lite (including our own).

**Comment by snej** [comment]
> it breaks binary compatibility because the method signature changes

What method?

**Comment by borrrden** [comment]
CryptoStream CreateStream(Stream baseStream) -> Stream CreateStream(Stream baseStream); (The .NET equivalent of encryptorBlock in Objective-C)

and [here](https://github.com/APXLabs/couchbase-lite-net/pull/3/commits/cab381cdd974d4bac602e9f502c4ff184920f8ae) in various places switching from SymmetricKey to ISymmetricKey.

**Comment by borrrden** [comment]
@elevine I haven't forgotten about this issue, it is being consulted as we move forward towards our 2.0 API

**Comment by snej** [comment]
I've just finished re-implementing attachment storage for the 2.0 core engine, and changed the encryption to a format that's seekable.

Source code is [here](https://github.com/couchbase/couchbase-lite-core/blob/master/LiteCore/Support/EncryptedStream.cc). Basically, the blob is written in independent 4k chunks. I'm not using CTR mode because I had trouble getting it to work correctly on iOS/Mac; it's still CTR, but the IV is the chunk number, which I believe is functionally equivalent. The hardest part was handling data that's not a whole number of chunks long. I ended up using no padding for the full chunks, and PKCS7 for the final incomplete chunk. If data _is_ a whole number of chunks long, I have to append a zero-length chunk (with padding).

Code/security review is welcome. Or if you think your format is better, let me know how it works.
